### PR TITLE
Fix SynchronousCircuitBreaker - reset counters correctly

### LINKED ISSUE #28: Circuitbreaker not opening with certain parameters with fallback method
Hi,
I am trying to use the `CircuitBreaker` annotation with following parameters
```
@Fallback(fallbackMethod = "handleLoginFailure")
@CircuitBreaker(successThreshold = 1,requestVolumeThreshold = 2, failureRatio = 0.5,
delay = 15, failOn = {Exception.class}, delayUnit = ChronoUnit.SECONDS)
```
I make two sucessfull requests as expected and then it breaks as expected on the first unsuccessful request, but the subsequent request passes through as if it was closed. Am I missing something in the configuration or is this expected behaviour? Should'nt the delay be active now and the fallback method should be handling the subsequent requests during the duration of the delay?

/David

### PR DESCRIPTION
**Pull request by mkouba**
- i.e. whenever the requestVolumeThreshold is reached, even if the
status did not change
- resolves #28

### CODE REVIEW

**Comment by mkouba** [comment]
In fact, I'm not really sure this PR is correct. I'm still confused. `CircuitBreaker#requestVolumeThreshold()` mentions a "rolling window". Does it mean we should always check the last `x` requests? And reset the window only if the circuit status changes?

**Comment by mkouba** [comment]
In fact, I'm not really sure this PR is correct. I'm still confused. `CircuitBreaker#requestVolumeThreshold()` mentions a "rolling window". Does it mean we should always check the last `x` requests? And reset the window only if the circuit status changes?

**Comment by mkouba** [comment]
@antoinesd pls review this PR carefully.

I've also created two related spec issues:
https://github.com/eclipse/microprofile-fault-tolerance/issues/342
https://github.com/eclipse/microprofile-fault-tolerance/issues/341

File: implementation/src/main/java/io/smallrye/faulttolerance/SynchronousCircuitBreaker.java [hunk 1]
```diff
@@ -190,13 +199,27 @@ private boolean isHalfOpenAttemptAllowed() {
...
+    private void reset() {
+        successCount.set(0);
+        halfOpenAttempts.set(0);
+        rollingWindow.clear();
     }
 
-    private void resetCounts() {
-        successCount.set(0);
-        failureCount.set(0);
+    private int getFailureCount() {
```
Comment (michalszynkiewicz) [inline implementation/src/main/java/io/smallrye/faulttolerance/SynchronousCircuitBreaker.java]: It's probably practical to keep the window small so going through the lit shouldn't be a problem but if we wanted to be faster we could additionally keep counters as we did, just update them also based on the removed value.

Comment (mkouba) [reply implementation/src/main/java/io/smallrye/faulttolerance/SynchronousCircuitBreaker.java]: @michalszynkiewicz Yes, you're right. I think that a window will usually contain dozens of requests which shouldn't be a problem. Also I was thinking that in future we might need to store other attributes of a request, such as timestamp.

Comment (mkouba) [reply implementation/src/main/java/io/smallrye/faulttolerance/SynchronousCircuitBreaker.java]: By the way, how would you distinguish which result to remove if we only have counters? I.e. no ordering info.

Comment (michalszynkiewicz) [reply implementation/src/main/java/io/smallrye/faulttolerance/SynchronousCircuitBreaker.java]: I meant the list that we have now + counters

Comment (mkouba) [reply implementation/src/main/java/io/smallrye/faulttolerance/SynchronousCircuitBreaker.java]: I see!

File: implementation/src/main/java/io/smallrye/faulttolerance/SynchronousCircuitBreaker.java [hunk 2]
```diff
@@ -207,10 +230,11 @@ private void resetCounts() {
 
     private final AtomicInteger successCount;
 
-    private final AtomicInteger failureCount;
-
     private final AtomicInteger halfOpenAttempts;
 
+    // failure = true, success = false
+    private final LinkedList<Boolean> rollingWindow;
```
Comment (michalszynkiewicz) [inline implementation/src/main/java/io/smallrye/faulttolerance/SynchronousCircuitBreaker.java]: Won't it misbehave with multiple threads? I think `ConcurrentLinkedQueue` should fit and be thread-safe.

File: testsuite/basic/src/test/java/io/smallrye/faulttolerance/circuitbreaker/rollingwindow/CircuitBreakerRollingWindowTest.java [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+        } catch (IllegalStateException expected) {
+        }
+        try {
+            pingService.ping(true);
+            fail("Circuit should be open now");
+        } catch (CircuitBreakerOpenException expected) {
+        }
+        assertEquals(6, pingService.getPingCounter().get());
+
+        Thread.sleep(300);
```
Comment (michalszynkiewicz) [inline testsuite/basic/src/test/java/io/smallrye/faulttolerance/circuitbreaker/rollingwindow/CircuitBreakerRollingWindowTest.java]: what's the role of `sleep` here?

File: implementation/src/main/java/io/smallrye/faulttolerance/SynchronousCircuitBreaker.java (hunk 2 above)
Comment (mkouba) [reply implementation/src/main/java/io/smallrye/faulttolerance/SynchronousCircuitBreaker.java]: Well, it's only manipulated from within synchronized methods. But `ConcurrentLinkedDeque` might be a better fit.

File: testsuite/basic/src/test/java/io/smallrye/faulttolerance/circuitbreaker/rollingwindow/CircuitBreakerRollingWindowTest.java (hunk 3 above)
Comment (mkouba) [reply testsuite/basic/src/test/java/io/smallrye/faulttolerance/circuitbreaker/rollingwindow/CircuitBreakerRollingWindowTest.java]: `@CircuitBreaker(delay = 300)`

File: implementation/src/main/java/io/smallrye/faulttolerance/SynchronousCircuitBreaker.java (hunk 2 above)
Comment (Ladicek) [reply implementation/src/main/java/io/smallrye/faulttolerance/SynchronousCircuitBreaker.java]: Doesn't the rolling window have a fixed size? Array-backed ring buffer should perform better than a linked list.

Comment (mkouba) [reply implementation/src/main/java/io/smallrye/faulttolerance/SynchronousCircuitBreaker.java]: It has and it should. But I don't think it makes sense to save few ms here.

Comment (Ladicek) [reply implementation/src/main/java/io/smallrye/faulttolerance/SynchronousCircuitBreaker.java]: The difference probably isn't even in millis :-) I actually think that plain `ArrayList` outperforms a `LinkedList` in almost all cases, but don't really mind.

**Comment by mkouba** [comment]
@antoinesd I think we're ready to merge ;-)

**Pull request merged**